processor/src/decoder/tests.rs

```diff
@@ -0,0 +1,675 @@
+
+    check_op_decoding(&trace, 0, ZERO, Operation::Loop, 0, 0, 0);
+    check_op_decoding(&trace, 1, INIT_ADDR, Operation::Span, 8, 0, 0);
+    check_op_decoding(&trace, 2, iter1_addr, Operation::Pad, 0, 0, 1);
+    check_op_decoding(&trace, 3, iter1_addr, Operation::Drop, 0, 1, 1);
+    check_op_decoding(&trace, 4, iter1_addr, Operation::End, 0, 0, 0);
+    // start second iteration
+    check_op_decoding(&trace, 5, INIT_ADDR, Operation::Repeat, 0, 0, 0);
+    check_op_decoding(&trace, 6, INIT_ADDR, Operation::Span, 8, 0, 0);
+    check_op_decoding(&trace, 7, iter2_addr, Operation::Pad, 0, 0, 1);
+    check_op_decoding(&trace, 8, iter2_addr, Operation::Drop, 0, 1, 1);
+    check_op_decoding(&trace, 9, iter2_addr, Operation::End, 0, 0, 0);
+    check_op_decoding(&trace, 10, INIT_ADDR, Operation::End, 0, 0, 0);
+    check_op_decoding(&trace, 11, ZERO, Operation::Halt, 0, 0, 0);
+
+    // --- check hasher state columns -------------------------------------------------------------
+
+    // in the first row, the hasher state is set to the hash of the loop's body
+    let loop_body_hash: Word = loop_body.hash().into();
+    assert_eq!(loop_body_hash, get_hasher_state1(&trace, 0));
+    assert_eq!([ZERO; 4], get_hasher_state2(&trace, 0));
+
+    // at the end of the first iteration, the hasher state is also set to the hash of the loops
+    // body, and is_loop_body flag is also set to ONE
+    assert_eq!(loop_body_hash, get_hasher_state1(&trace, 4));
+    assert_eq!([ONE, ZERO, ZERO, ZERO], get_hasher_state2(&trace, 4));
+
+    // at the RESPAN row hasher state is copied over from the previous row
+    assert_eq!(loop_body_hash, get_hasher_state1(&trace, 5));
+    assert_eq!([ONE, ZERO, ZERO, ZERO], get_hasher_state2(&trace, 5));
+
+    // at the end of the second iteration, the hasher state is again set to the hash of the loops
+    // body, and is_loop_body flag is also set to ONE
+    assert_eq!(loop_body_hash, get_hasher_state1(&trace, 9));
+    assert_eq!([ONE, ZERO, ZERO, ZERO], get_hasher_state2(&trace, 9));
+
+    // the hash of the program is located in the last END row; this row should also have is_loop
+    // flag set to ONE
+    let program_hash: Word = program.hash().into();
+    assert_eq!(program_hash, get_hasher_state1(&trace, 10));
+    assert_eq!([ZERO, ONE, ZERO, ZERO], get_hasher_state2(&trace, 10));
+
+    // HALT opcode and program hash gets propagated to the last row
+    for i in 12..trace_len {
+        assert!(contains_op(&trace, i, Operation::Halt));
+        assert_eq!(program_hash, get_hasher_state1(&trace, i));
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+fn build_trace(stack: &[u64], program: &CodeBlock) -> (DecoderTrace, usize) {
+    let inputs = ProgramInputs::new(stack, &[], vec![]).unwrap();
+    let mut process = Process::new(inputs);
+    process.execute_code_block(&program).unwrap();
```

Comment:
nit: I don't think we need this borrow. It could just be `program`.

Reply:
Fixed!